### PR TITLE
Check for location permissions when loading Nearby

### LINKED ISSUE #650: Ask for location permissions at runtime for Nearby Places
While you use Nearby feature in new beta it doesnt show any nearby items; warning "No places nearby is shown instead. GPS on; Android 6.0

### PR DESCRIPTION
**Pull request by misaochan**
Fix #650 . Adds runtime permissions checking when Nearby is loaded, tells user that Nearby places cannot be found if permissions denied.

### CODE REVIEW

**Comment by misaochan** [comment]
I disagree with Codacy that I should be splitting a 101 character line up any further. >.> Not sure if we want a default clause for the permissions result switch?

**Comment by tobias47n9e** [comment]
I can confirm for emulator API 24 and 25 that it asks me for the permissions. Denying the permissions goes to an endless loading screen. Maybe a message "You need GPS permissions to use Nearby" would be more helpful than the loading icon. When the permissions are granted I still sometimes get no results, probably because of https://github.com/commons-app/apps-android-commons/issues/674.

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivity.java [hunk 1]
```diff
@@ -89,6 +90,44 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+                ActivityCompat.requestPermissions(this,
+                        new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, LOCATION_REQUEST);
+            }
+        } else {
+            startLookingForNearby();
+        }
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
```
Comment (tobias47n9e) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivity.java]: Android Studio complains here that the superclass method uses `@NonNull`.

**Comment by misaochan** [comment]
Ah, good point. I'm not sure how useful it would be to have a screen with just those words though. Maybe I could use the toast, but just send them back to the previous activity.

**Comment by tobias47n9e** [comment]
Just for reference this is what I am thinking of:

screenshot_2017-05-31-20-01-47

screenshot_2017-05-31-19-59-17

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivity.java (hunk 1 above)
Comment (misaochan) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivity.java]: Good catch, thanks!

**Comment by misaochan** [comment]
The user should now be taken to a screen that explains Nearby Places can't be shown if they deny permission. It's not anywhere near as pretty as the screenshots here :) , but that could be a future enhancement. Also, a future enhancement could give the user an option to be asked for permissions again at NoPermissionsFragment without having to go back to NearbyActivity to do it, but I don't think that is essential for the time being.

Note that I have included a temporary fix for #709 . I considered doing a separate PR for that, but since that fix is absolutely needed in order for this PR to work, it made sense to me to tie them together.

File: app/build.gradle [hunk 2]
```diff
@@ -64,4 +64,8 @@ android {
         disable 'ExtraTranslation'
         abortOnError false
     }
+
+    configurations.all {
+        resolutionStrategy.force 'com.android.support:support-annotations:25.2.0'
+    }
```
Comment (whym) [inline app/build.gradle]: I suggest leaving a comment ("FIXME" or "TODO") saying this is temporary or pointing to the issue.

Comment (misaochan) [inline app/build.gradle]: Good point, added

**Comment by tobias47n9e** [comment]
@misaochan I really like the way you implemented the message!

I am justing testing the branch. If I deny location permissions then the nearby mapview will crash in the `showMapView()` function in the `NearbyActivity`. Haven't been able to figure out why yet.

**Comment by misaochan** [comment]
Thanks @tobias47n9e . :) Could you post the crash logs?

**Comment by tobias47n9e** [comment]
```
06-02 19:07:17.736 3376-3376/fr.free.nrw.commons E/AndroidRuntime: FATAL EXCEPTION: main
Process: fr.free.nrw.commons, PID: 3376
java.lang.NullPointerException:
Attempt to invoke virtual method 'android.os.AsyncTask$Status fr.free.nrw.commons.nearby.NearbyActivity$NearbyAsyncTask.getStatus()' on a null object reference
at fr.free.nrw.commons.nearby.NearbyActivity.showMapView(NearbyActivity.java:180)
at fr.free.nrw.commons.nearby.NearbyActivity.onOptionsItemSelected(NearbyActivity.java:82)
at android.app.Activity.onMenuItemSelected(Activity.java:3204)
at android.support.v4.app.FragmentActivity.onMenuItemSelected(FragmentActivity.java:408)
at android.support.v7.app.AppCompatActivity.onMenuItemSelected(AppCompatActivity.java:195)
at android.support.v7.view.WindowCallbackWrapper.onMenuItemSelected(WindowCallbackWrapper.java:113)
at android.support.v7.view.WindowCallbackWrapper.onMenuItemSelected(WindowCallbackWrapper.java:113)
at android.support.v7.app.ToolbarActionBar$2.onMenuItemClick(ToolbarActionBar.java:69)
at android.support.v7.widget.Toolbar$1.onMenuItemClick(Toolbar.java:206)
at android.support.v7.widget.ActionMenuView$MenuBuilderCallback.onMenuItemSelected(ActionMenuView.java:776)
at android.support.v7.view.menu.MenuBuilder.dispatchMenuItemSelected(MenuBuilder.java:822)
at android.support.v7.view.menu.MenuItemImpl.invoke(MenuItemImpl.java:156)
at android.support.v7.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:969)
at android.support.v7.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:959)
at android.support.v7.widget.ActionMenuView.invokeItem(ActionMenuView.java:623)
at android.support.v7.view.menu.ActionMenuItemView.onClick(ActionMenuItemView.java:154)
at android.view.View.performClick(View.java:5610)
at android.view.View$PerformClick.run(View.java:22265)
at android.os.Handler.handleCallback(Handler.java:751)
at android.os.Handler.dispatchMessage(Handler.java:95)
at android.os.Looper.loop(Looper.java:154)
at android.app.ActivityThread.main(ActivityThread.java:6077)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
```

**Comment by misaochan** [comment]
The recent commit should fix it hopefully.

**Comment by tobias47n9e** [comment]
@misaochan Clicking the map button works now avoids the crash. Just one small problem remains. Pressing refresh in that view makes the message disappear and after that I can click back and forth on the map without any messages anymore that I am missing the location permissions.

**Comment by misaochan** [comment]
If that is the only remaining issue and it otherwise works fine, shall we merge first and make that a separate issue (possibly a beginner friendly enhancement)? I was hoping to push this fix to beta by Monday, as currently everyone in beta has to give permissions manually. :)

**Pull request merged**